lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ldap/JdkLdapClient.java

```diff
@@ -97,10 +100,41 @@ public <T> T processLdapContext(String userName, String password, LdapContextPro
     public <T> T executeLdapQuery(String userName, String password, LdapQuery ldapQuery, LdapSearchResultProcessor<T> resultProcessor)
             throws NamingException
     {
-        try (CloseableContext context = createUserDirContext(userName, password);
-                CloseableSearchResults search = searchContext(ldapQuery, context)) {
-            return resultProcessor.process(search.searchResults);
+        final int maxRetry = 2;
+        for (int retry = 0; retry <= maxRetry; retry++) {
+            CloseableContext context = ctxs.get(userName);
+            if (context == null) {
+                // Ensure only one context was created per user.
+                synchronized (ctxs) {
+                    context = ctxs.get(userName);
+                    if (context == null) {
```

Comment:
hmm...not sure why we check context == null twice. maybe just

  context = createUserDirContext(userName, password);
  ctxs.put(userName, context);
is enough?

Reply:
Because the first comparison was on the value fetched without lock, so there is a chance that multiple threads reach the synchronize clause at the same time.